### PR TITLE
menu: improve menu generation management

### PR DESCRIPTION
**Pull request by zannkukai**
Improves the menu generation management. Adding some specifications on
menu entries :
  * allowing to set disabled state on any menu entry.
  * allowing to separate entries with a divider.
  * removing some css indentation from the menu component.
Removes the 'PrefixSuffixComponent' and replaces it with a simple
template.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## How to test?

- use the ng-core-tester to test the menu component : `npm run serve`

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 1]
```diff
@@ -95,7 +124,7 @@ export class MenuItem implements MenuItemInterface {
    * @return boolean
    */
   hasFactory() {
-    return this._factory ? true : false;
+    return !!this._factory;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:127]: Note : this notation has proposed by my editor and seems working well !

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 2]
```diff
@@ -68,6 +67,36 @@ export class MenuItem implements MenuItemInterface {
...
+  private _enabled = true;
+
+
+  /**
+   * Delete
+   * @param name - string, name of attribute
+   * @param attributes - dictionnary of attributes
+   * @return boolean
+   */
+  private static _deleteAttr(name: string, attributes: {}): boolean {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:80]: Make private function at the end of file

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 3]
```diff
@@ -68,6 +67,36 @@ export class MenuItem implements MenuItemInterface {
...
+    return false;
+  }
+
+  /**
+   * Set Prefix Suffix
+   * @param name - string
+   * @param htmlClasses - null or string
+   * @return Object with name and class attributes
+   */
+  private static _setPrefixSuffix(name: string, htmlClasses: null | string) {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:93]: Make private function at the end of file

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 4]
```diff
@@ -571,49 +599,37 @@ export class MenuItem implements MenuItemInterface {
...
-   * @param dict - dictionnary of attributes
-   * @return boolean
+   * enable the menuItem
+   * @return MenuItem
    */
-  private has(name: string, dict = {}): boolean {
-    return Object.keys(dict).find(element => element === name)
-      ? true
-      : false;
+  enable(): this {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:613]: Delete this. Comment above

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 5]
```diff
@@ -571,49 +599,37 @@ export class MenuItem implements MenuItemInterface {
...
+   * enable the menuItem
+   * @return MenuItem
    */
-  private _deleteAttr(name: string, attributes: {}): boolean {
-    if (this.has(name, attributes)) {
-      delete attributes[name];
-      return true;
-    }
-    return false;
+  disable(): this {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:622]: Delete this. Comment above

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html [hunk 6]
```diff
@@ -1,110 +1,99 @@
...
-              <ng-core-menu-widget-prefix-suffix [item]="item"></ng-core-menu-widget-prefix-suffix>
-            </span>
+            <ng-container [ngTemplateOutlet]="childMenuEntry" [ngTemplateOutletContext]="{item:item}"></ng-container>
           </a>
           <ng-template #linkRouterLink>
             <a
-              *ngIf="!item.isActive(); else menuNoLink"
               class="nav-link"
+              [class.active]="item.isActive()"
+              [class.disabled]="!item.enabled()"
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:24]: `[class.disabled]="!item.isEnabled()"`

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html [hunk 7]
```diff
@@ -1,110 +1,99 @@
...
+                class="dropdown-menu"
+                [ngClass]="item.getAttribute('class')"
+                role="menu"
+                *dropdownMenu
+            >
+              <ng-container *ngFor="let child of item.getChildren()">
+                <li>
+                  <a *ngIf="child.hasUri(); else subLinkRouterLink"
+                     class="dropdown-item"
+                     [class.disabled]="!child.enabled()"
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:53]: `[class.disabled]="!child.isEnabled()"`

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html [hunk 8]
```diff
@@ -1,110 +1,99 @@
...
+                     [attr.id]="child.getAttribute('id')"
+                     [attr.href]="child.getUri()"
+                     [attr.target]="child.getAttribute('target')"
+                     (click)="doClick(child)"
+                  >
+                    <ng-container [ngTemplateOutlet]="childMenuEntry" [ngTemplateOutletContext]="{item:child}"></ng-container>
+                  </a>
+                  <ng-template #subLinkRouterLink>
+                    <a class="dropdown-item"
+                       [class.disabled]="!child.enabled()"
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:63]: `[class.disabled]="!child.isEnabled()"`

File: projects/rero/ng-core/src/lib/menu/menu-item.ts [hunk 9]
```diff
@@ -571,49 +599,37 @@ export class MenuItem implements MenuItemInterface {
...
 
   /**
-   * Count
-   * @return integer, number of children
+   * Is the menuItem is enabled
+   * @return boolean
    */
-  count(): number {
-    return this._children.length;
+  enabled(): boolean {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-item.ts:605]: Why not make `isEnabled` and delete `enable` and `disable` function ?

**Review by iGormilhit — COMMENTED** [review]
Commit message proposition:

```
menu: improve menu generation management

Improves the menu generation management by adding some specifications on
menu entries:
  * Allows to set disabled state on any menu entry.
  * Allows to separate entries with a divider.
  * Removes some CSS indentation from the menu component.
Removes the 'PrefixSuffixComponent' and replaces it with a simple
template.
```

File: projects/rero/ng-core/src/lib/menu/menu-item.ts (hunk 9 above)
Comment (zannkukai) [reply projects/rero/ng-core/src/lib/menu/menu-item.ts:605]: I change `enabled()` to `isEnabled()` to be consistant with others interface functions.
I will keep the `enable()` and `disable()` functions as setter for the private `_enabled` attribute.

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html [hunk 10]
```diff
@@ -1,110 +1,99 @@
...
-                </a>
-              </ng-template>
-            </ng-container>
-          </div>
         </li>
+        <ng-template #withDropdown>
+          <li class="nav-item dropdown" dropdown placement="bottom right">
+            <a class="nav-link dropdown-toggle"
+               dropdownToggle
+               [routerLink]=""
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:38]: Why empty `routerLink` ?

**Review by Garfield-fr — APPROVED** [review]

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html (hunk 10 above)
Comment (zannkukai) [reply projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:38]: Without this, the "a" tag isn't recognize as a valid link by browser and the cursor stay as pointer (not the little hand).

Comment (Garfield-fr) [reply projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:38]: and with `href="#"` ?

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

File: projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html (hunk 10 above)
Comment (sebastiendeleze) [reply projects/rero/ng-core/src/lib/menu/menu-widget/menu-widget.component.html:38]: I think simply adding `href` alone works.

**Review by sebastiendeleze — APPROVED** [review]

**Pull request merged**